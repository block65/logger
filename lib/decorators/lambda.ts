import { Decorator } from '../logger.js';

export const lambdaDecorator: Decorator = (log) => {
  return {
    ...log,
  };
};
