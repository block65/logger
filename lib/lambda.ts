import type { Context } from 'aws-lambda';
import type { JsonObject } from 'type-fest';
import type { Logger } from './logger.js';
import { isEmptyObject } from './utils.js';

type LambdaContext = Partial<Context> & Required<Pick<Context, 'awsRequestId'>>;

export function withLambdaLoggerContextWrapper<T>(
  logger: Logger,
  lambdaContext: LambdaContext,
  fn: (...args: any) => Promise<T>,
): Promise<T> {
  const context: JsonObject = {
    // functionName: context.functionName,
    ...(lambdaContext.functionVersion !== '$LATEST' && {
      functionVersion: lambdaContext.functionVersion,
    }),
  };

  return logger.als.run(
    {
      id: lambdaContext.awsRequestId,
      ...(!isEmptyObject(context) && { context }),
    },
    () => {
      return fn();
    },
  );
}
