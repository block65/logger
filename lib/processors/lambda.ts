import { Processor } from '../logger.js';

export const lambdaProcessor: Processor = (log) => {
  return {
    ...log,
  };
};
