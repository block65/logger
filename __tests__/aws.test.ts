import {
  afterAll,
  afterEach,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { lambdaLoggerContextWrapper } from '../lib/lambda.js';
import { createLoggerWithWaitableMock } from './helpers.js';

describe('AWS', () => {
  beforeEach(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2009-02-13T23:31:30.000Z'));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('Lambda Logger + contextId', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();

    const contextId = 'fake-aws-request-id';

    const closure = lambdaLoggerContextWrapper(logger, contextId, {
      stuff: true,
    });

    expect.assertions(1);

    await new Promise<void>((resolve) => {
      closure(async () => {
        logger.warn('hello');
        await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
        resolve();
      });
    });
  });

  test('Lambda Error Object', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock({
      platform: 'aws-lambda',
    });

    logger.error(new Error('Ded 4'));
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Error Object', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock({
      platform: 'aws',
    });

    logger.error(new Error('Ded 3'));
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });
});
