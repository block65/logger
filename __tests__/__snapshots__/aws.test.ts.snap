// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AWS Error Object 1`] = `
Array [
  Object {
    "err": Object {
      "message": "Ded 3",
      "name": "Error",
      "stack": "Error: Ded 3
    at Object.<anonymous> (<rootDir>/__tests__/aws.test.ts:37:18)
    at Promise.then.completed (<rootDir>/node_modules/jest-circus/build/utils.js:390:28)
    at new Promise (<anonymous>)",
    },
    "hostname": "testhost",
    "level": 50,
    "msg": "Ded 3",
    "pid": 1337,
    "time": 1234567890000,
  },
]
`;

exports[`AWS Lambda Error Object 1`] = `
Array [
  Object {
    "err": Object {
      "message": "Ded 4",
      "name": "Error",
      "stack": "Error: Ded 4
    at Object.<anonymous> (<rootDir>/__tests__/aws.test.ts:28:18)
    at Promise.then.completed (<rootDir>/node_modules/jest-circus/build/utils.js:390:28)
    at new Promise (<anonymous>)",
    },
    "level": 50,
    "msg": "Ded 4",
    "time": 1234567890000,
  },
]
`;

exports[`AWS Lambda Logger + contextId 1`] = `
Array [
  Object {
    "_context": Object {
      "stuff": true,
    },
    "_contextId": "fake-aws-request-id",
    "hostname": "testhost",
    "level": 40,
    "msg": "hello",
    "pid": 1337,
    "time": 1234567890000,
  },
]
`;
